src/bench/checkblock.cpp

```diff
@@ -50,5 +54,43 @@ static void DeserializeAndCheckBlockTest(benchmark::Bench& bench)
     });
 }
 
+static void SerializeNetMsg(benchmark::Bench& bench)
+{
+    CDataStream stream(benchmark::data::block413567, SER_NETWORK, PROTOCOL_VERSION);
+    std::byte a{0};
+    stream.write({&a, 1});
+    CBlock block;
+    stream >> block;
+    auto pcmpctblock = std::make_shared<const CBlockHeaderAndShortTxIDs>(block);
+    const CNetMsgMaker msgMaker(PROTOCOL_VERSION);
+    bench.run([&] {
+        CSerializedNetMsg msg{msgMaker.Make(NetMsgType::CMPCTBLOCK, *pcmpctblock)};
+        ankerl::nanobench::doNotOptimizeAway(msg);
+    });
+}
+
+static void CacheSerializeNetMsg(benchmark::Bench& bench)
+{
+    CDataStream stream(benchmark::data::block413567, SER_NETWORK, PROTOCOL_VERSION);
+    std::byte a{0};
+    stream.write({&a, 1});
+    CBlock block;
+    stream >> block;
+    auto pcmpctblock = std::make_unique<const CBlockHeaderAndShortTxIDs>(block);
+    const std::shared_future<CSerializedNetMsg> lazy_ser{
+        std::async(std::launch::deferred, [pcmpctblock = move(pcmpctblock)] {
+            const CNetMsgMaker msgMaker(PROTOCOL_VERSION);
+            return msgMaker.Make(NetMsgType::CMPCTBLOCK, *pcmpctblock);
+        }
+    )};
+
+    bench.run([&] {
+        CSerializedNetMsg msg{lazy_ser.get().Copy()};
```

Comment:
Not sure how useful this is. Won't this massively elevate the duration of the first call compared to all following ones?

The speed-up I reported in https://github.com/bitcoin/bitcoin/pull/23880#issuecomment-1086043016 is obviously dependent on the number of peers you send the message to. If there is only one peer, it will be slower, because the duration is `Ser+Copy*1`, which is more work than just `Ser*1`. However with sufficient peers `N`, there is a speed-up because `Ser+Copy*N` is faster than `Ser*N`.

Reply:
Hmm I suppose a better benchmark would be time it takes to copy to see what `N` needs to be for this optimization to make sense. I was making the assumption here that `Copy` is always faster than `Ser`, so if there's at least one high bandwidth peer we relayed too already then it makes sense to just `Copy` for any low bandwidth requests that come in rather than `Ser`.

Reply:
I think the confusion is that this PR was changing the first low bandwidth response to use a lazy serializer and copy. This is actually using a copy of the same lazy future that is used for the high bandwidth peers. So as long as there's at least one high bandwidth peer then this optimization makes sense because the serialization will already have been done.